File: src/main/java/br/unifor/costify/domain/valueobject/RecipeIngredient.java
package br.unifor.costify.domain.valueobject;

import br.unifor.costify.domain.errors.recipe.InvalidQuantityException;
import java.util.Objects;

public class RecipeIngredient {
  private final Id ingredientId;
  private final double quantity;
  private final Unit unit;

  public RecipeIngredient(Id ingredientId, double quantity, Unit unit) {
    if (quantity <= 0) {
      throw new InvalidQuantityException("Quantity must be greater than zero");
    }

    this.ingredientId = ingredientId;
    this.quantity = quantity;
    this.unit = unit;
  }

  public Id getIngredientId() {
    return ingredientId;
  }

  public double getQuantity() {
    return quantity;
  }

  public Unit getUnit() {
    return unit;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    RecipeIngredient that = (RecipeIngredient) o;
    return Double.compare(quantity, that.quantity) == 0
        && Objects.equals(ingredientId, that.ingredientId)
        && unit == that.unit;
  }

  @Override
  public int hashCode() {
    return Objects.hash(ingredientId, quantity, unit);
  }
}

File: internal/domain/valueobject/recipe_ingredient.go
package valueobject

import (
	"github.com/vini/costify-go/internal/domain/errors"
)

// RecipeIngredient represents an ingredient used in a recipe with quantity and unit
type RecipeIngredient struct {
	ingredientId Id
	quantity     float64
	unit         Unit
}

// NewRecipeIngredient creates a new RecipeIngredient with validation
func NewRecipeIngredient(ingredientId Id, quantity float64, unit Unit) (RecipeIngredient, error) {
	if quantity <= 0 {
		return RecipeIngredient{}, errors.NewInvalidQuantityError("Quantity must be greater than zero")
	}

	return RecipeIngredient{
		ingredientId: ingredientId,
		quantity:     quantity,
		unit:         unit,
	}, nil
}

// IngredientId returns the ingredient ID
func (ri RecipeIngredient) IngredientId() Id {
	return ri.ingredientId
}

// Quantity returns the quantity
func (ri RecipeIngredient) Quantity() float64 {
	return ri.quantity
}

// Unit returns the unit
func (ri RecipeIngredient) Unit() Unit {
	return ri.unit
}

// Equals checks if two RecipeIngredients are equal
func (ri RecipeIngredient) Equals(other RecipeIngredient) bool {
	return ri.ingredientId.Equals(other.ingredientId) &&
		ri.quantity == other.quantity &&
		ri.unit == other.unit
}
